File: 12. Error Handling and Building Fault Tolerant Systems/code/typescript/go_deep_health_check_endpoint_2.ts
// TypeScript, Deep health check endpoint
import express, { Request, Response } from 'express';
import { Pool } from 'pg';
import { Redis } from 'ioredis';

interface HealthStatus {
    status: string;
    checks: Record<string, string>;
}

export function healthHandler(db: Pool, rdb: Redis) {
    return async (req: Request, res: Response) => {
        const checks: Record<string, string> = {};
        let overall = "ok";

        // DB check (with basic timeout simulation via Promise.race)
        try {
            const dbCheck = db.query('SELECT 1');
            const timeout = new Promise((_, reject) => setTimeout(() => reject(new Error("Timeout")), 2000));
            await Promise.race([dbCheck, timeout]);
            checks["database"] = "ok";
        } catch (err: any) {
            checks["database"] = "unhealthy: " + err.message;
            overall = "degraded";
        }

        // Redis check
        try {
            await rdb.ping();
            checks["cache"] = "ok";
        } catch (err: any) {
            checks["cache"] = "unhealthy: " + err.message;
            overall = "degraded";
        }

        const status = overall !== "ok" ? 503 : 200;

        res.setHeader("Content-Type", "application/json");
        res.status(status).json({ status: overall, checks });
    };
}

File: 12. Error Handling and Building Fault Tolerant Systems/code/rust/go_deep_health_check_endpoint_2.rs
// Rust, Deep health check endpoint
use actix_web::{web, HttpResponse, Responder};
use sqlx::PgPool;
use redis::aio::ConnectionManager;
use serde::Serialize;
use std::collections::HashMap;
use std::time::Duration;
use tokio::time::timeout;

#[derive(Serialize)]
pub struct HealthStatus {
    pub status: String,
    pub checks: HashMap<String, String>,
}

pub async fn health_handler(
    db: web::Data<PgPool>,
    rdb: web::Data<ConnectionManager>
) -> impl Responder {
    let mut checks = HashMap::new();
    let mut overall = "ok".to_string();

    // DB check
    let db_ping = async { sqlx::query("SELECT 1").execute(db.get_ref()).await };
    match timeout(Duration::from_secs(2), db_ping).await {
        Ok(Ok(_)) => {
            checks.insert("database".to_string(), "ok".to_string());
        }
        Ok(Err(e)) => {
            checks.insert("database".to_string(), format!("unhealthy: {}", e));
            overall = "degraded".to_string();
        }
        Err(e) => {
            checks.insert("database".to_string(), format!("unhealthy: {}", e));
            overall = "degraded".to_string();
        }
    }

    // Redis check
    let mut conn = rdb.get_ref().clone();
    match redis::cmd("PING").query_async::<_, String>(&mut conn).await {
        Ok(_) => {
            checks.insert("cache".to_string(), "ok".to_string());
        }
        Err(e) => {
            checks.insert("cache".to_string(), format!("unhealthy: {}", e));
            overall = "degraded".to_string();
        }
    }

    let status = if overall != "ok" {
        actix_web::http::StatusCode::SERVICE_UNAVAILABLE
    } else {
        actix_web::http::StatusCode::OK
    };

    HttpResponse::build(status).json(HealthStatus { status: overall, checks })
}
